File: src/Sample/Plumix.Sample/Demos/General/KeyboardListenerDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

namespace Plumix;

// Dart parity source: dart_sample/lib/demos/general/keyboard_listener_demo_page.dart (exact sample parity)

public sealed class KeyboardListenerDemoPage : StatefulWidget
{
    public override State CreateState()
    {
        return new KeyboardListenerDemoPageState();
    }
}

internal sealed class KeyboardListenerDemoPageState : State
{
    private readonly FocusNode _keyboardFocusNode = new();
    private readonly FocusNode _rawFocusNode = new();
    private readonly FocusNode _shortcutFocusNode = new();
    private readonly FocusNode _excludedFocusNode = new();
    private IReadOnlyDictionary<ShortcutActivator, Intent> _shortcuts = null!;
    private IReadOnlyDictionary<Type, FlutterAction> _actions = null!;
    private string _keyboardEvent = "none";
    private string _rawEvent = "none";
    private int _shortcutCount;
    private int _excludedClickCount;
    private bool _shortcutFocusHighlight;
    private bool _shortcutHoverHighlight;

    public override void InitState()
    {
        _keyboardFocusNode.AddListener(HandleFocusChanged);
        _rawFocusNode.AddListener(HandleFocusChanged);
        _shortcutFocusNode.AddListener(HandleFocusChanged);
        _excludedFocusNode.AddListener(HandleFocusChanged);
        _shortcuts = new Dictionary<ShortcutActivator, Intent>
        {
            [new SingleActivator("K", control: true)] = new CounterShortcutIntent(1),
            [new SingleActivator("J", control: true)] = new CounterShortcutIntent(-1),
        };
        _actions = new Dictionary<Type, FlutterAction>
        {
            [typeof(CounterShortcutIntent)] = new CallbackAction<CounterShortcutIntent>(
                intent =>
                {
                    SetState(() => _shortcutCount += intent.Delta);
                    return _shortcutCount;
                })
        };
    }

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12,
            children:
            [
                new Text("Keyboard + focus action detectors", fontSize: 20, color: Colors.Black),
                new Text(
                    "Click a panel or use its button, then press and release keyboard keys.",
                    fontSize: 14,
                    color: Colors.DimGray),
                BuildKeyboardListenerProbe(),
                BuildRawKeyboardListenerProbe(),
                BuildActionsShortcutsProbe(),
            ]);
    }

    public override void Dispose()
    {
        _keyboardFocusNode.RemoveListener(HandleFocusChanged);
        _rawFocusNode.RemoveListener(HandleFocusChanged);
        _shortcutFocusNode.RemoveListener(HandleFocusChanged);
        _excludedFocusNode.RemoveListener(HandleFocusChanged);
        _keyboardFocusNode.Dispose();
        _rawFocusNode.Dispose();
        _shortcutFocusNode.Dispose();
        _excludedFocusNode.Dispose();
    }

    private Widget BuildKeyboardListenerProbe()
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 8,
            children:
            [
                new TextButton(
                    onPressed: () => _keyboardFocusNode.RequestFocus(),
                    child: new Text("Focus KeyboardListener")),
                new KeyboardListener(
                    focusNode: _keyboardFocusNode,
                    onKeyEvent: keyEvent => SetState(() =>
                    {
                        _keyboardEvent = DescribeKeyEvent(keyEvent);
                    }),
                    child: BuildPanel(
                        title: "KeyboardListener",
                        detail: _keyboardEvent,
                        focused: _keyboardFocusNode.HasFocus)),
            ]);
    }

    private Widget BuildRawKeyboardListenerProbe()
    {
#pragma warning disable CS0618
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 8,
            children:
            [
                new TextButton(
                    onPressed: () => _rawFocusNode.RequestFocus(),
                    child: new Text("Focus RawKeyboardListener")),
                new RawKeyboardListener(
                    focusNode: _rawFocusNode,
                    onKey: keyEvent => SetState(() =>
                    {
                        string phase = keyEvent is RawKeyDownEvent ? "down" : "up";
                        _rawEvent = $"{keyEvent.Key} — {phase}";
                    }),
                    child: BuildPanel(
                        title: "RawKeyboardListener (deprecated compatibility)",
                        detail: _rawEvent,
                        focused: _rawFocusNode.HasFocus)),
            ]);
#pragma warning restore CS0618
    }

    private Widget BuildActionsShortcutsProbe()
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 8,
            children:
            [
                new TextButton(
                    onPressed: () => _shortcutFocusNode.RequestFocus(),
                    child: new Text("Focus Actions + Shortcuts")),
                new FocusableActionDetector(
                    focusNode: _shortcutFocusNode,
                    shortcuts: _shortcuts,
                    actions: _actions,
                    onShowFocusHighlight: value => SetState(() => _shortcutFocusHighlight = value),
                    onShowHoverHighlight: value => SetState(() => _shortcutHoverHighlight = value),
                    child: BuildPanel(
                        title: "FocusableActionDetector",
                        detail:
                        $"count {_shortcutCount} — Ctrl+K / Ctrl+J — focus highlight "
                        + $"{OnOff(_shortcutFocusHighlight)} — hover {OnOff(_shortcutHoverHighlight)}",
                        focused: _shortcutFocusNode.HasFocus)),
                new ExcludeFocusTraversal(
                    child: new TextButton(
                        focusNode: _excludedFocusNode,
                        onPressed: () => SetState(() => _excludedClickCount++),
                        child: new Text(
                            $"ExcludeFocusTraversal: Tab skips, click works ({_excludedClickCount})"))),
            ]);
    }

    private static Widget BuildPanel(string title, string detail, bool focused)
    {
        return new Container(
            height: 88,
            padding: new Thickness(12),
            decoration: new BoxDecoration(
                Color: focused ? Color.Parse("#FFE0F2F1") : Color.Parse("#FFF1F3F4"),
                Border: new BorderSide(
                    color: focused ? Color.Parse("#FF00796B") : Color.Parse("#FF9AA0A6"),
                    width: focused ? 2 : 1),
                BorderRadius: BorderRadius.Circular(10)),
            child: new Column(
                mainAxisAlignment: MainAxisAlignment.Center,
                crossAxisAlignment: CrossAxisAlignment.Start,
                spacing: 6,
                children:
                [
                    new Text(title, fontSize: 15, color: Colors.Black),
                    new Text($"Last event: {detail}", fontSize: 13, color: Colors.DimGray),
                ]));
    }

    private static string DescribeKeyEvent(KeyEvent keyEvent)
    {
        return $"{keyEvent.Key} — {(keyEvent.IsDown ? "down" : "up")}";
    }

    private static string OnOff(bool value) => value ? "on" : "off";

    private void HandleFocusChanged()
    {
        SetState(static () => { });
    }
}

internal sealed class CounterShortcutIntent : Intent
{
    public CounterShortcutIntent(int delta)
    {
        Delta = delta;
    }

    public int Delta { get; }
}

File: dart_sample/lib/demos/general/keyboard_listener_demo_page.dart
// ignore_for_file: deprecated_member_use

import 'package:material_ui/material_ui.dart';
import 'package:flutter/services.dart';

class KeyboardListenerDemoPage extends StatefulWidget {
  const KeyboardListenerDemoPage({super.key});

  @override
  State<KeyboardListenerDemoPage> createState() =>
      _KeyboardListenerDemoPageState();
}

class _KeyboardListenerDemoPageState extends State<KeyboardListenerDemoPage> {
  final FocusNode _keyboardFocusNode = FocusNode();
  final FocusNode _rawFocusNode = FocusNode();
  final FocusNode _shortcutFocusNode = FocusNode();
  final FocusNode _excludedFocusNode = FocusNode();
  late final Map<ShortcutActivator, Intent> _shortcuts;
  late final Map<Type, Action<Intent>> _actions;
  String _keyboardEvent = 'none';
  String _rawEvent = 'none';
  int _shortcutCount = 0;
  int _excludedClickCount = 0;
  bool _shortcutFocusHighlight = false;
  bool _shortcutHoverHighlight = false;

  @override
  void initState() {
    super.initState();
    _keyboardFocusNode.addListener(_handleFocusChanged);
    _rawFocusNode.addListener(_handleFocusChanged);
    _shortcutFocusNode.addListener(_handleFocusChanged);
    _excludedFocusNode.addListener(_handleFocusChanged);
    _shortcuts = <ShortcutActivator, Intent>{
      const SingleActivator(LogicalKeyboardKey.keyK, control: true):
          const _CounterShortcutIntent(1),
      const SingleActivator(LogicalKeyboardKey.keyJ, control: true):
          const _CounterShortcutIntent(-1),
    };
    _actions = <Type, Action<Intent>>{
      _CounterShortcutIntent: CallbackAction<_CounterShortcutIntent>(
        onInvoke: (_CounterShortcutIntent intent) {
          setState(() {
            _shortcutCount += intent.delta;
          });
          return _shortcutCount;
        },
      ),
    };
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'Keyboard + focus action detectors',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Click a panel or use its button, then press and release keyboard keys.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        _buildKeyboardListenerProbe(),
        _buildRawKeyboardListenerProbe(),
        _buildActionsShortcutsProbe(),
      ],
    );
  }

  @override
  void dispose() {
    _keyboardFocusNode.removeListener(_handleFocusChanged);
    _rawFocusNode.removeListener(_handleFocusChanged);
    _shortcutFocusNode.removeListener(_handleFocusChanged);
    _excludedFocusNode.removeListener(_handleFocusChanged);
    _keyboardFocusNode.dispose();
    _rawFocusNode.dispose();
    _shortcutFocusNode.dispose();
    _excludedFocusNode.dispose();
    super.dispose();
  }

  Widget _buildKeyboardListenerProbe() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 8,
      children: <Widget>[
        TextButton(
          onPressed: _keyboardFocusNode.requestFocus,
          child: const Text('Focus KeyboardListener'),
        ),
        KeyboardListener(
          focusNode: _keyboardFocusNode,
          onKeyEvent: (KeyEvent event) {
            setState(() {
              _keyboardEvent = _describeKeyEvent(event);
            });
          },
          child: _buildPanel(
            title: 'KeyboardListener',
            detail: _keyboardEvent,
            focused: _keyboardFocusNode.hasFocus,
          ),
        ),
      ],
    );
  }

  Widget _buildRawKeyboardListenerProbe() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 8,
      children: <Widget>[
        TextButton(
          onPressed: _rawFocusNode.requestFocus,
          child: const Text('Focus RawKeyboardListener'),
        ),
        RawKeyboardListener(
          focusNode: _rawFocusNode,
          onKey: (RawKeyEvent event) {
            setState(() {
              final String phase = event is RawKeyDownEvent ? 'down' : 'up';
              _rawEvent = '${event.logicalKey.debugName} — $phase';
            });
          },
          child: _buildPanel(
            title: 'RawKeyboardListener (deprecated compatibility)',
            detail: _rawEvent,
            focused: _rawFocusNode.hasFocus,
          ),
        ),
      ],
    );
  }

  Widget _buildActionsShortcutsProbe() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 8,
      children: <Widget>[
        TextButton(
          onPressed: _shortcutFocusNode.requestFocus,
          child: const Text('Focus Actions + Shortcuts'),
        ),
        FocusableActionDetector(
          focusNode: _shortcutFocusNode,
          shortcuts: _shortcuts,
          actions: _actions,
          onShowFocusHighlight: (bool value) {
            setState(() {
              _shortcutFocusHighlight = value;
            });
          },
          onShowHoverHighlight: (bool value) {
            setState(() {
              _shortcutHoverHighlight = value;
            });
          },
          child: _buildPanel(
            title: 'FocusableActionDetector',
            detail:
                'count $_shortcutCount — Ctrl+K / Ctrl+J — focus highlight '
                '${_onOff(_shortcutFocusHighlight)} — hover '
                '${_onOff(_shortcutHoverHighlight)}',
            focused: _shortcutFocusNode.hasFocus,
          ),
        ),
        ExcludeFocusTraversal(
          child: TextButton(
            focusNode: _excludedFocusNode,
            onPressed: () {
              setState(() {
                _excludedClickCount++;
              });
            },
            child: Text(
              'ExcludeFocusTraversal: Tab skips, click works '
              '($_excludedClickCount)',
            ),
          ),
        ),
      ],
    );
  }

  static Widget _buildPanel({
    required String title,
    required String detail,
    required bool focused,
  }) {
    return Container(
      height: 88,
      padding: const EdgeInsets.all(12),
      decoration: BoxDecoration(
        color: focused ? const Color(0xFFE0F2F1) : const Color(0xFFF1F3F4),
        border: Border.all(
          color: focused ? const Color(0xFF00796B) : const Color(0xFF9AA0A6),
          width: focused ? 2 : 1,
        ),
        borderRadius: BorderRadius.circular(10),
      ),
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.start,
        spacing: 6,
        children: <Widget>[
          Text(
            title,
            style: const TextStyle(fontSize: 15, color: Colors.black),
          ),
          Text(
            'Last event: $detail',
            style: const TextStyle(fontSize: 13, color: Colors.black54),
          ),
        ],
      ),
    );
  }

  static String _describeKeyEvent(KeyEvent event) {
    final String phase = switch (event) {
      KeyDownEvent() => 'down',
      KeyRepeatEvent() => 'repeat',
      _ => 'up',
    };
    return '${event.logicalKey.debugName} — $phase — '
        'physical ${event.physicalKey.debugName}';
  }

  static String _onOff(bool value) => value ? 'on' : 'off';

  void _handleFocusChanged() {
    setState(() {});
  }
}

class _CounterShortcutIntent extends Intent {
  const _CounterShortcutIntent(this.delta);

  final int delta;
}
